### PR TITLE
Volume control & draggable sliders

### PR DESCRIPTION
**Pull request by kkuepper**

### CODE REVIEW

**Review by andreasgangso — CHANGES_REQUESTED** [review]
didnt read detailed but id refactor into a reusable slider component instead of embedding it inside the mediaplayer.

also consider https://zagjs.com/components/vue-sfc/slider for some free functionality and accessibility (zag is modular, pnpm install this and import)

**Comment by kkuepper** [comment]
I looked at different sliders but didn't like them for different reasons. But the slider from zag looks really good. I'll refactor to use it.

File: components/media-player/MediaPlayerOpen.vue [hunk 1]
```diff
@@ -24,22 +26,68 @@ const {
...
+  volume,
 } = useNuxtApp().$mediaPlayer;
 
-const onPointerUpProgressBar = (event: PointerEvent) => {
-  const rect = (event.currentTarget as Element)?.getBoundingClientRect();
-  currentPosition.value =
-    ((event.clientX - rect.left) / rect.width) * currentTrackDuration.value;
+const [state, send] = useMachine(
+  slider.machine({
+    id: "volume",
```
Comment (kkuepper) [inline components/media-player/MediaPlayerOpen.vue:34]: @andreasgangso I refactored the volume slider. Do you by chance know how to make a 2nd slider work?
I haven't really tried it much but my first try caused some problems.

Comment (kkuepper) [reply components/media-player/MediaPlayerOpen.vue]: done

File: components/media-player/PositionSlider.vue [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+const currentOrNewPosition = computed(() =>
+  newPosition.value === null
+    ? currentPosition.value
+    : (newPosition.value / 100) * currentTrackDuration.value,
+);
+
+watch([currentPosition, currentTrackDuration], () => {
+  if (newPosition.value === null) {
+    const value = (currentPosition.value / currentTrackDuration.value) * 100;
```
Comment (andreasgangso) [inline components/media-player/PositionSlider.vue:38]: is division by zero possible?

File: plugins/mediaPlayer/mediaPlayer.ts [hunk 3]
```diff
@@ -193,6 +197,16 @@ export const initMediaPlayer = (
     },
   });
 
+  const volumeComputed = computed({
```
Comment (andreasgangso) [inline plugins/mediaPlayer/mediaPlayer.ts:200]: id prefer exposing volume and doing:
```ts
watch(volume, (v) => {
  activeMedia.value?.setVolume(v);
});
```

File: components/media-player/PositionSlider.vue [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    id: "position",
+    value: [0],
+    step: 0.01,
+    onValueChange(details) {
+      const [value] = details.value;
+      if (value) newPosition.value = value;
+    },
+    onValueChangeEnd(details) {
+      const [value] = details.value;
+      if (value) {
```
Comment (andreasgangso) [inline components/media-player/PositionSlider.vue:19]: why the if (value)? is it nullable?
0 is falsy, so maybe you want to do `value != null` (null or undefined) instead?

File: components/media-player/VolumeSlider.vue [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+const { volume } = useNuxtApp().$mediaPlayer;
+
+const [volumeState, volumeSend] = useMachine(
+  slider.machine({
+    id: "volume",
+    value: [volume.value * 100],
+    onValueChange(details) {
+      const [value] = details.value;
+      if (value) volume.value = value / 100;
```
Comment (andreasgangso) [inline components/media-player/VolumeSlider.vue:13]: if (value) ?

**Review by andreasgangso — CHANGES_REQUESTED** [review]
some comments

File: plugins/mediaPlayer/mediaPlayer.ts (hunk 3 above)
Comment (kkuepper) [reply plugins/mediaPlayer/mediaPlayer.ts:200]: I think the current solution is more in line with what is already being done in the mediaPlayer

File: components/media-player/PositionSlider.vue (hunk 4 above)
Comment (kkuepper) [reply components/media-player/PositionSlider.vue:19]: zag-js defines the value as number | undefined

File: components/media-player/PositionSlider.vue (hunk 2 above)
Comment (kkuepper) [reply components/media-player/PositionSlider.vue:38]: changed it

**Pull request merged**